### ISSUE TITLE
Improving project symbol search

### ISSUE THREAD

**Issue by joeldrapper** [zed-industries/zed#13307]
### Check for existing issues

- [X] Completed

### Describe the feature

There seems to be a difference between the way buffer symbols and project symbols are generated. I think buffer symbols use Tree-sitter while project symbols use the language server? Buffer symbols seem to be much faster and more accurate and reliable.

Would it makes sense to unify these and have Tree-sitter generate project symbols too? Additionally, project symbols seem to be generated lazily. Sometimes you need to open the relevant file before the symbols become available.

Would it be possible to configure Zed to build and maintain an index of project symbols up-front? I appreciate the performance and memory usage considerations, but I have 128GB of memory and would _love_ to be able to search my whole project and jump straight to the relevant symbol.

**Comment by maxdeviant** [comment]
I don’t think using Tree-sitter for project symbol search makes sense, as it means we’d have to open each file in the project, parse it with Tree-sitter, and then collect up the results.

While we could use an index to speed this up, it does seem like we’re replicating language server features at that point.

Whereas the language server already has a way of introspecting the project for searchable symbols.

**Comment by IkerAriz** [comment]
I'd find a treesitter-based global search very useful, especially in mixed-language setups. For instance, a project that includes a C++ library with a function `X`, a corresponding python binding module for that C++ library, and a mix of C++ and Python code that calls `X`. A more specific search for any `X` call sites, regardless of language, could offer a nice improvement over grep-based search. I imagine this could be accomplished with sufficiently configured language servers, but a useful symbol search that works out of the box would be a pretty good option (especially when warming up to a new project).

**Comment by joeldrapper** [comment]
I would imagine a project-wide tree-sitter symbol index would be very useful for AI too. AI tools could use multi-step prompts where the AI can drive, searching for symbols and pulling in context.

**Comment by mitranim** [comment]
The lack of this feature is a deal-breaker for me. LSPs don't exist for many languages (which I use). Currently I use Sublime Text, which does support global indexing, and it's indispensable.

**Comment by joeldrapper** [comment]
It’s not a deal breaker for me but it would be incredibly useful. The tree-sitter outlines are so much more reliable than a lot of language servers.

**Comment by mitranim** [comment]
Worth noting that not only many languages lack a language server, but also that LS-based references don't work across languages. Which is a shame, because in multi-language projects, you often end up with many cross-language references embedded in comments. Like in JS client code, you'd refer to a server Go type, where you'd refer to an SQL table, where you'd refer to a server API function, and so on. These kinds of references are very useful, and work seamlessly in editors with universal indexing (such as Sublime Text's "goto definition").

**Comment by auwi-nordic** [comment]
I second this request, it's definitely a dealbreaker. I work on a big multi-language project. Sublime Text just works. VS Code kinda works (slow, but possible to configure it to be acceptable).

I think there should be an option to use this for "Go to Definition" and "Find all references" as well. It wouldn't be as precise as LSP, but as long as you get a list of matches to choose from for the definition it should be more robust than LSP for some use-cases.

**Comment by joeldrapper** [comment]
Yes! And we know it works because the outline view when you’re looking at a specific file is absolutely brilliant. We just need the ability to search all the outlines for all files.

**Comment by olejorgenb** [comment]
Another improvement to the symbols search would be to allow filtering on filepath.

<img width="551" height="295" alt="Image" src="https://github.com/user-attachments/assets/a06dcd5c-22c5-4756-a583-8bc516025c72" />

<img width="555" height="99" alt="Image" src="https://github.com/user-attachments/assets/2edc3b7a-1277-47a0-bd61-e9cf9e7b0d4c" />

^^ imaging the result being filtered down to the object.rs match ^^

Perhaps using some special marker would be best `SYMBOL NAME QUERY / FILE PATH QUERY`. The reverse order could be done by `/ FILE PATH QUERY @ SYMBOL NAME QUERY`

**Comment by ahdyt** [comment]
Hi is this issue related to something like Go, jump to usage?

e.g.

func main() {
   hello() // clicking for here jump to hello definition fine
}

func hello() { // clicking hello() cant jump to hello() inside main() sometimes, the problem is in single file, if it's in different file, it's fine. The cursor just doesn't jump sometimes.
}
